pkg/controllers/cluster/cluster_controller.go

```diff
@@ -51,15 +67,81 @@ type Controller struct {
 	ClusterMonitorGracePeriod time.Duration
 	// When cluster is just created, e.g. agent bootstrap or cluster join, we give a longer grace period.
 	ClusterStartupGracePeriod time.Duration
+	// FailoverEvictionTimeout represents the grace period for deleting scheduling result on failed clusters.
+	FailoverEvictionTimeout time.Duration
 
 	// Per Cluster map stores last observed health together with a local time when it was observed.
-	clusterHealthMap sync.Map
+	clusterHealthMap *clusterHealthMap
+}
+
+// NewController returns a new cluster controller.
+func NewController(
```

Comment:
Yes, this time we need a constructor because we have some private field to be initialized here. In `cluster_controller`, `clusterHealthMap` is a private map which need to be created in this constructor.

Reply:
I get it. But it's not the only way we can achieve that, such as initialize it in `SetupWithManager()`.

Reply:
That's right. Which one is better?

Reply:
I'd prefer the option with `SetupWithManager()`, I know some company not allowed to have a function with more than 5 parameters. And I don't want to extend the function each time when we introduce a new option to it.

Reply:
OK, thanks. this is my last concern, otherwise LGTM.